src/main/java/org/ethereum/beacon/discovery/storage/LocalNodeRecordStore.java

```diff
@@ -21,33 +22,43 @@ public LocalNodeRecordStore(
       final SecretKey secretKey,
       final NodeRecordListener recordListener,
       final NewAddressHandler newAddressHandler) {
-    this.latestRecord = record;
+    this.latestRecord = new AtomicReference<>(record);
     this.secretKey = secretKey;
     this.recordListener = recordListener;
     this.newAddressHandler = newAddressHandler;
   }
 
   public NodeRecord getLocalNodeRecord() {
-    return latestRecord;
+    return latestRecord.get();
   }
 
   public void onSocketAddressChanged(final InetSocketAddress newAddress) {
-    NodeRecord oldRecord = this.latestRecord;
     newAddressHandler
-        .newAddress(oldRecord, newAddress)
+        .newAddress(latestRecord.get(), newAddress)
         .ifPresent(
-            record -> {
-              this.latestRecord = record;
-              if (!record.equals(oldRecord)) {
-                recordListener.recordUpdated(oldRecord, record);
+            newRecord -> {
+              final NodeRecord oldRecord = latestRecord.getAndSet(newRecord);
```

Comment:
Does it still look as non-atomic update here? `get` and `getAndSet` are not under lock and the value can be modified between these calls.

Reply:
Haven't looked into this again. Busy getting PeerDAS into master :)

Will have a second look soon!